src/libraries/Common/src/System/Net/IPv6AddressHelper.Common.cs

```diff
@@ -376,40 +408,48 @@ internal static void Parse(ReadOnlySpan<char> address, Span<ushort> numbers, int
                     case '/':
                         if (numberIsValid)
                         {
-                            numbers[index++] = (ushort)number;
+                            numbers[index++] = number;
                             numberIsValid = false;
                         }
 
-                        // since we have a valid IPv6 address string, the prefix
-                        // length is the last token in the string
-                        for (++i; address[i] != ']'; ++i)
+                        for (++i; i < address.Length && address[i] != TChar.CreateTruncating(']'); i++)
                         {
-                            PrefixLength = PrefixLength * 10 + (address[i] - '0');
                         }
+
                         break;
 
                     default:
-                        number = number * 16 + Uri.FromHex(address[i++]);
+                        TChar ch = address[i++];
+                        int castCharacter = int.CreateTruncating(ch);
+                        int characterValue = castCharacter switch
+                        {
+                            >= '0' and <= '9' => castCharacter - '0',
+                            >= 'a' and <= 'f' => castCharacter - 'a' + 10,
+                            >= 'A' and <= 'F' => castCharacter - 'A' + 10,
+                            _ => 0x0
+                        };
```

Comment:
HexConvert.FromChar accepts an int. Can we just use that here and in other places where you've open-coded the hex comparisons?